athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -0,0 +1,215 @@
+            try {
+                // Create a new search-source injected with the projection, predicate, and the pagination batch size.
+                SearchSourceBuilder searchSource = new SearchSourceBuilder().size(QUERY_BATCH_SIZE)
+                        .fetchSource(ElasticsearchHelper
+                                .getProjection(recordsRequest.getSchema())).query(ElasticsearchHelper
+                                .getQuery(recordsRequest.getConstraints().getSummary()));
+                // Create a new search-request for the specified index.
+                SearchRequest searchRequest = new SearchRequest(recordsRequest.getTableName().getTableName());
+                int hitsNum;
+                int currPosition = 0;
+                do {
+                    // Process the search request injecting the search-source, and setting the from position
+                    // used for pagination of results.
+                    SearchResponse searchResponse = client
+                            .getDocuments(searchRequest.source(searchSource.from(currPosition)));
+                    // Increment current position to next batch of results.
+                    currPosition += QUERY_BATCH_SIZE;
+                    // Process hits.
+                    Iterator<SearchHit> hitIterator = searchResponse.getHits().iterator();
+                    hitsNum = searchResponse.getHits().getHits().length;
+
+                    while (hitIterator.hasNext() && queryStatusChecker.isQueryRunning()) {
+                        ++numRows;
+                        processDocument(spiller, recordsRequest.getSchema().getFields().iterator(),
+                                client.getDocument(hitIterator.next()), numResultRows);
+                    }
+                    // if hitsNum < QUERY_BATCH_SIZE, then this is the last batch of documents.
+                } while (hitsNum == QUERY_BATCH_SIZE && queryStatusChecker.isQueryRunning());
+            }
+            catch (IOException error) {
+                logger.error("Error sending query:", error);
+            }
+            finally {
+                client.shutdown();
+            }
+        }
+
+        logger.info("readWithConstraint: numRows[{}] numResultRows[{}]", numRows, numResultRows.get());
+    }
+
+    /**
+     * Process the Document extracted from the SearchHit and write it's data into the BlockSpiller.
+     * @param spiller is a BlockSpiller that should be used to write the row data associated with this Document.
+     *                The BlockSpiller automatically handles chunking the response, encrypting, and spilling to S3.
+     * @param fieldIterator is the iterator for the list of schema fields extracted from the RecordRequest.
+     * @param document is the Document to be processed.
+     * @param numResultRows is the number of successfully processed Rows (Documents).
+     */
+    protected void processDocument(BlockSpiller spiller, Iterator<Field> fieldIterator,
```

Comment:
We should move to the extractor framework here. Its significantly faster (4x in most cases than this approach of using lots of branching and conditionals for field extraction and coercion.

Reply:
Replaced with field extractors.